File: course/section_4_stack/code/13-validate-stack-sequences.py
from typing import List


def pr[T](val: T, title="") -> T:
    # print(title, val)
    return val


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        if pushed == popped:
            return True
        i = 0
        stack = []
        for num in pushed:
            stack.append(num)
            while stack and i < len(popped) and stack[-1] == popped[i]:
                stack.pop()
                i += 1
        return not stack

File: course/section_4_stack/code/13-validate-stack-sequences.php
<?php
class Solution {

    /**
     * @param Integer[] $pushed
     * @param Integer[] $popped
     * @return Boolean
     */
    function validateStackSequences($pushed, $popped) {
        $stack = [] ;
        $i = 0 ; 
        foreach($pushed as $num){
            $stack[] = $num ; 
            while(!empty($stack) && $i < count($popped) && $stack[count($stack)-1] == $popped[$i]){
                array_pop($stack);
                $i++;
            }
        }
        return empty($stack);
        
    }
}